### PR TITLE
Feat/decouple buttons

### LINKED ISSUE #492: Hot reloading server
https://github.com/gaearon/react-transform-boilerplate
https://github.com/yetu/controlcenter
https://github.com/chicoxyzzy/react-for-hipsters-boilerplate/

### LINKED ISSUE #503: Investigate karma with karma-webpack for tests
Ideally with HMR integaration

### LINKED ISSUE #524: Rewrite Docs Entries to React
Rewrite docs-block and all related things to React

### LINKED ISSUE #526: Add basic tests setup
Add basic tests setup for react components and ES6 modules

### LINKED ISSUE #578: jsx return expressions look sloppy
Using round brackets can help.
``` jsx
return (
  <div>
    <Component />
  </div>
);
```
*vs*
``` jsx
 return <div>
     <Component />
 </div>
```

### PR DESCRIPTION
**Pull request by mr-mig**
The scope of work:

- I've splitted buttons and refactored them to `React.createClass` to have true HMR (fixed #492).
- Test setup is `karma + jasmine` instead of `karma + mocha + chai + sinon`. Added custom matcher for checking class existence in generated DOM output.
- The full TDD setup is ready and working as expected (fixed #503, fixed #526).
- Docs entries moved to React (fixed #524)
- Decouple button styles
- Proper ES6 syntax for react return syntax and arrow functions (fixed #578)
- Fixed scss-lint checks, added tests to travic (ref #535)

In the next PR:
- Write tests for all button modifiers (ref #526)
- Implement icons (needed for FB button) (ref #577 ref #550)

Test output with `karma-mocha-reporter` looks like that: 
image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,16 +31,16 @@
...
     "gulp-util": "^3.0.6",
     "husky": "^0.10.1",
+    "jasmine-core": "^2.3.4",
     "json-loader": "^0.5.3",
     "karma": "^0.13.15",
     "karma-chrome-launcher": "^0.2.1",
     "karma-cli": "^0.1.1",
-    "karma-mocha": "^0.2.0",
-    "karma-sinon-chai": "^1.1.0",
+    "karma-jasmine": "^0.2.2",
```
Comment (mr-mig) [inline package.json]: Basically, `jasmine === mocha + chai + sinon` with less verbose syntax and easier custom matchers

File: src/components/component-base/hole/hole.spec.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import Hole from './index';
+import $ from 'teaspoon';
+
+describe('hole', function () {
+  it('should render class with __hole element', function () {
+    let result = $(<Hole className="test"/>).render();
```
Comment (mr-mig) [inline src/components/component-base/hole/hole.spec.js]: This is a `teaspoon` library. It mimics jquery for React components.
To be used in tests.

File: src/components/mint-button-primary/mint-button-primary.spec.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  it('should have dark type available', () => {
+    expect(types.dark).toBeDefined();
+  });
+
+  it('should have alt type available', () => {
+    expect(types.alt).toBeDefined();
+  });
+
+  it('should have mint-button-primary class', () => {
+    expect($(<Button/>).render()).toHaveClass('mint-button-primary');
```
Comment (mr-mig) [inline src/components/mint-button-primary/mint-button-primary.spec.js]: The output should match BEM structure we have right now.
This should be checked in tests for all components.

File: src/components/mint-button-secondary/index.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import klassy from 'helpers/klassy';
+import Hole from 'components/component-base/hole';
+
+export const availableTypes = {
+  dark: 'dark', alt: 'alt'
+};
+
+const classes = klassy(pj);
+
+export default React.createClass({
```
Comment (mr-mig) [inline src/components/mint-button-secondary/index.js]: Buttons should be generalized. TBD.

File: src/components/mint-button-secondary/mint-button-secondary.scss [hunk 5]
```diff
@@ -1,4 +1,4 @@
-@import '~sass/main.scss';
+@import '~sass/main';
```
Comment (mr-mig) [inline src/components/mint-button-secondary/mint-button-secondary.scss]: This thing breaks Travis integration (scss-lint).
There's an issue to fix that: #535

File: src/docs/components/docs/index.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  });
+
+}
+
+export default (props) => {
+  return <DocsEntry title="Buttons">
+    <DocsBlock title="Primary Buttons">
+
+      <MintButton>Primary</MintButton>
+      { renderAll(btnTypes, 'type', MintButton) }
```
Comment (mr-mig) [inline src/docs/components/docs/index.js]: Render all buttons with all available modifiers for `type`

File: tools/tests.webpack.js [hunk 7]
```diff
@@ -1,2 +1,9 @@
+var matchers = require('helpers/teaspoon-jasmine-matchers');
```
Comment (mr-mig) [inline tools/tests.webpack.js]: Somehow this class cannot use ES6 modules :(

File: src/components/mint-button-primary/mint-button-primary.scss [hunk 8]
```diff
@@ -0,0 +1,312 @@
+@import '~sass/main.scss';
```
Comment (mr-mig) [inline src/components/mint-button-primary/mint-button-primary.scss]: This scss should be generalized for 2 components. 
Right now the file is copied twice.
TBD

File: tools/tests.webpack.js [hunk 9]
```diff
@@ -1,2 +1,9 @@
...
+console.log(matchers);
```
Comment (iurii-kondratiuk) [inline tools/tests.webpack.js]: Is there any purpose to keep this `console.log` ?

Comment (mr-mig) [inline tools/tests.webpack.js]: Not really, will remove it, thx for spotting.

File: package.json [hunk 10]
```diff
@@ -31,16 +31,16 @@
...
+    "karma-mocha-reporter": "^1.1.1",
```
Comment (aju) [inline package.json]: Why we need mocha reporter when you are using jasmine?

File: src/components/mint-button-primary/index.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import './mint-button-primary.scss';
+import pj from './package.json';
+import klassy from 'helpers/klassy';
+import Hole from 'components/component-base/hole';
+
+export const availableTypes = {
```
Comment (aju) [inline src/components/mint-button-primary/index.js]: Maybe we should have object config that is returned with options inside?

File: package.json (hunk 10 above)
Comment (mr-mig) [inline package.json]: The output is nicer :)
Pure aesthetic reasons

File: src/components/mint-button-primary/index.js (hunk 11 above)
Comment (mr-mig) [inline src/components/mint-button-primary/index.js]: Can you give me an example, plz?

File: src/components/mint-button-secondary/mint-button-secondary.spec.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+import React from 'react';
+import { availableTypes as types, default as Component } from './index';
+import $ from 'teaspoon';
+
+describe('mint-button-secondary', () => {
+  it('should have class mint-button-secondary', () => {
+    expect($(<Component/>).render()).toHaveClass('mint-button-secondary');
+  });
+
+  it('should have class mint-button--alt', () => {
```
Comment (mirrh) [inline src/components/mint-button-secondary/mint-button-secondary.spec.js]: should have class mint-button-**secondary**--alt

File: src/components/mint-button-secondary/mint-button-secondary.spec.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    expect(types.alt).toBeDefined();
+    expect($(<Component type="alt"/>).render()).toHaveClass('mint-button-secondary--alt');
+  });
+
+  it('should have class mint-button--dark', () => {
```
Comment (mirrh) [inline src/components/mint-button-secondary/mint-button-secondary.spec.js]: and here

File: src/components/mint-button-primary/mint-button-primary.spec.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  });
+
+  it('should have mint-button--alt class', () => {
```
Comment (mirrh) [inline src/components/mint-button-primary/mint-button-primary.spec.js]: lost 'primary'
(should have mint-button-**primary**--alt class)

File: src/components/mint-button-primary/mint-button-primary.spec.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    expect($(<Button type="alt"/>).render()).toHaveClass('mint-button-primary--alt');
+  });
+
+  it('should have mint-button--dark class', () => {
```
Comment (mirrh) [inline src/components/mint-button-primary/mint-button-primary.spec.js]: and here

File: src/components/mint-button-secondary/mint-button-secondary.spec.js (hunk 12 above)
Comment (mr-mig) [inline src/components/mint-button-secondary/mint-button-secondary.spec.js]: good catch. 
All of those descriptions copy-pasted. Will fix it.

File: tools/tests.webpack.js (hunk 7 above)
Comment (dev-tim) [inline tools/tests.webpack.js]: Sure, because you define config for the webpack, that will define a loader, that will allow to use ES6.

Comment (mr-mig) [inline tools/tests.webpack.js]: There is a loader in karma config already. 
I do not get your proposition.

Comment (dev-tim) [inline tools/tests.webpack.js]: Ah it was just a suggestion

**Comment by dev-tim** [comment]
+1

**Pull request merged**